src/main/java/executor/service/service/impl/ScenarioExecutorImpl.java

```diff
@@ -1,12 +1,65 @@
 package executor.service.service.impl;
 
 import executor.service.model.Scenario;
+import executor.service.model.Step;
 import executor.service.service.ScenarioExecutor;
+import executor.service.service.StepExecutionClickCss;
+import executor.service.service.StepExecutionClickXpath;
+import executor.service.service.StepExecutionSleep;
+import org.openqa.selenium.NoSuchElementException;
 import org.openqa.selenium.WebDriver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+/**
+ * The {@code ScenarioExecutorImpl} class implements the {@link ScenarioExecutor} interface
+ * and selects a StepExecution for each Step and runs it.
+ *
+ * @author Oleksandr Tuleninov
+ * @version 01
+ * @see executor.service.service.impl.stepExecution.StepExecutionClickCssImpl
+ * @see executor.service.service.impl.stepExecution.StepExecutionSleepImpl
+ * @see executor.service.service.impl.stepExecution.StepExecutionClickXpathImpl
+ * @see executor.service.model.Scenario
+ * @see executor.service.model.Step
+ */
 public class ScenarioExecutorImpl implements ScenarioExecutor {
+
+    private static final Logger log = LoggerFactory.getLogger(ScenarioExecutorImpl.class);
+
+    private final StepExecutionClickCss stepExecutionClickCss;
+    private final StepExecutionSleep stepExecutionSleep;
+    private final StepExecutionClickXpath stepExecutionClickXpath;
+
+    public ScenarioExecutorImpl(StepExecutionClickCss stepExecutionClickCss,
+                                StepExecutionSleep stepExecutionSleep,
+                                StepExecutionClickXpath stepExecutionClickXpath) {
+        this.stepExecutionClickCss = stepExecutionClickCss;
+        this.stepExecutionSleep = stepExecutionSleep;
+        this.stepExecutionClickXpath = stepExecutionClickXpath;
+    }
+
     @Override
     public void execute(Scenario scenario, WebDriver webDriver) {
+        webDriver.get(scenario.getSite());
 
+        Step currentStep = null;
+        try {
+            for (Step step : scenario.getSteps()) {
+                currentStep = step;
+                String action = step.getAction();
+                switch (action) {
+                    case "clickCss" -> stepExecutionClickCss.step(webDriver, step);
+                    case "sleep" -> stepExecutionSleep.step(webDriver, step);
+                    case "clickXpath" -> stepExecutionClickXpath.step(webDriver, step);
+                    default -> log.error("Invalid step action: " + action);
+                }
+            }
+            log.info("Scenario: " + scenario.getName() + " is passed");
+        } catch (NoSuchElementException e) {
+            log.error("Scenario: " + scenario.getName() + " failed on Step: " + currentStep);
+        } finally {
+            webDriver.close();
+        }
```

Comment:
try-catch should be in loop, initialization or currentStep is redunant, it doesn't  throw exception,
Exception throw StepExecutors

there are no sense to set webdriver in finally block

Reply:
if try-catch will be in the loop, the execution will continue after step failure

Reply:
If you check that there are some values of css elements on the page without change path, and there are not only one css element, but all scenario would fail?

Reply:
Do you mean that there can be multiple elements with the same value on the same page, or that the following steps will not be executed after incorrect step?

Reply:
Yes, I tell you about that steps willn't be executed. If scenario consist only from clicking of css elements it's look's not very good idea, and even it failed on xpath, may be better show all failed checks, than broke all execution on one step? Also thought low probability, but sleep could be interrupted, broke all testing?

Reply:
> if try-catch will be in the loop, the execution will continue after step failure

then why dont you use *break*?

```
for (int i = 0; i < someArray.length; i++) {
    try {
        // Code that might throw an exception
        // ...
    } catch (SomeException e) {
        // Handle the exception if needed
        // ...
        
        // Use break to stop the loop
        break;
    }
}```

Reply:
> Yes, I tell you about that steps willn't be executed. If scenario consist only from clicking of css elements it's look's not very good idea, and even it failed on xpath, may be better show all failed checks, than broke all execution on one step? Also thought low probability, but sleep could be interrupted, broke all testing?

I don't quite get to the point. You say that, for example, in this scenario:
CLICK_CSS > SLEEP > CLICK_XPATH
execution should continue if CLICK_CSS fails?

Reply:
> > Yes, I tell you about that steps willn't be executed. If scenario consist only from clicking of css elements it's look's not very good idea, and even it failed on xpath, may be better show all failed checks, than broke all execution on one step? Also thought low probability, but sleep could be interrupted, broke all testing?
> 
> I don't quite get to the point. You say that, for example, in this scenario: CLICK_CSS > SLEEP > CLICK_XPATH execution should continue if CLICK_CSS fails?

Yes, why not? As I understood, CLICK_CSS check if there are  elements according to specified conditions.
CLICK_XPATH try to find link to address  on page and them forward to this address.

So there can be different situations, in one case when tests only css elements on single page, like this (plus if there are no exception) :
CLICK_CSS > + 
SLEEP         > +
CLICK_CSS > - 
SLEEP         > +  
CLICK_CSS > +
SLEEP         > +
CLICK_CSS > +


In this test result if you decide to use exception in loop client will see that there are problem only with one element, but no with another. 
So what better see one crash or all mistakes?

Otherwise if you look at this example, if  CLICK_XPATH is forward or redirect to another page would something like this:
CLICK_CSS > + 
SLEEP         > +
CLICK_XPATH > - 
SLEEP         > +  
CLICK_CSS > -
SLEEP         > +
CLICK_CSS > -

As I know, sometimes using click-xpath, there no redirect, but hide or appear some html elements and so on.

Reply:
The purpose of this application is control, as a consequence of the violation of the work of one Step already speaks of violations of the contract. The scenarios may be different, in them the Steps may or may not depend on each other. Our task is to create a universal approach for website testing. All the above facts make it possible to conclude that it is better to log at the Steps launch level, and not in Step. The final decision on this issue can be made by the customer. At the moment we have the following control result:
![image](https://github.com/GFLCourses5/worker/assets/140727960/c29f2c09-fac0-4b62-9ca7-f334fd21a6c3)

![image](https://github.com/GFLCourses5/worker/assets/140727960/3c1e1cf9-a739-46ee-a9dc-96b0b1d4659e)

Reply:
We always use click() on found element, so in our case the next CLICK_CSS should be on the other page. If step fails, there is no reason to try to execute steps on page that doesn't relate to scenario flow (on the wrong page). The client will see exatcly on which step execution failed. That means that previous steps was successful. And subsequent ones are meaningless.

Reply:
> We always use click() on found element, so in our case the next CLICK_CSS should be on the other page. If step fails, there is no reason to try to execute steps on page that doesn't relate to scenario flow (on the wrong page). The client will see exatcly on which step execution failed. That means that previous steps was successful. And subsequent ones are meaningless.

Why he is should be on the next page? We haven't any validator of scenario, only proxy, so if there all scenario consist only from Step with action sleep it would work. It can consist from all click-css elements. We haven't any words of the way execution.  We can have all steps that consists from all elements clickc-css or click-xpath. Or we will get crash if there be such scenario?

Reply:
> The final decision on this issue can be made by the customer.

It's very good answer, but is there can such way of execution about which I said? Can you answer yes or no? Please, stop avoiding direct answers.

Reply:
I will say more, perhaps a number of other situations that need to be foreseen, and about which we do not know.

Reply:
> Why he is should be on the next page? We haven't any validator of scenario, only proxy, so if there all scenario consist only from Step with action sleep it would work. It can consist from all click-css elements. We haven't any words of the way execution. We can have all steps that consists from all elements clickc-css or click-xpath. Or we will get crash if there be such scenario?

With given requirements, the only action performed is to click the element. We don't have requirements to implement actions like sendKeys, so it is reasonable to click the link and check what's on next page. Of course, we can simply check if element with given selector exists, but than why click on it? Even then, execution will continue to completion if all elements are present.

Execution will stop exactly on the step that failed. For example, scenario has 5 CLICK_CSS steps, and the third one is incorrect. In this case execution will stop on the third step, and this step will be reported. No reason to go further, because previous conditions are not satisfied.

Reply:
> I will say more, perhaps a number of other situations that need to be foreseen, and about which we do not know.

So right now you already have bean know about it. Is time to think about ways of solution?

Reply:
> > Why he is should be on the next page? We haven't any validator of scenario, only proxy, so if there all scenario consist only from Step with action sleep it would work. It can consist from all click-css elements. We haven't any words of the way execution. We can have all steps that consists from all elements clickc-css or click-xpath. Or we will get crash if there be such scenario?
> 
> With given requirements, the only action performed is to click the element. We don't have requirements to implement actions like sendKeys, so it is reasonable to click the link and check what's on next page. Of course, we can simply check if element with given selector exists, but than why click on it? Even then, execution will continue to completion if all elements are present.
> 
> Execution will stop exactly on the step that failed. For example, scenario has 5 CLICK_CSS steps, and the third one is incorrect. In this case execution will stop on the third step, and this step will be reported. No reason to go further, because previous conditions are not satisfied.

Okey, let's go from another side, your decide to test some values of css elements on the one page. Is it more convenient for you to get a detailed answer about where something went wrong or to make call to service  every time when you fix one bug?